src/twisted/logger/_format.py

```diff
@@ -287,3 +238,179 @@ def formatWithCall(formatString, mapping):
+    Each defaults to C{u"-"} is not set.  If formatting fails completely,
+    "UNFORMATTABLE" is returned.
+
+    @param event: The event containing the system specification.
+    @type event: L{dict}
+
+    @return: A formatted string representing the "log_system" key.
+    @rtype: L{unicode}
+    """
+    system = event.get("log_system", None)
+    if system is None:
+        level = event.get("log_level", None)
+        if level is None:
+            levelName = u"-"
+        else:
+            levelName = level.name
+
+        system = u"{namespace}#{level}".format(
+            namespace=event.get("log_namespace", u"-"),
+            level=levelName,
+        )
+    else:
+        try:
+            system = unicode(system)
+        except Exception:
+            system = u"UNFORMATTABLE"
+    return system
+
+
+
+def eventAsText(
+        event,
+        includeTraceback=True,
+        includeTimestamp=True,
+        includeSystem=True,
+        formatTime=formatTime,
+):
+    """
+    Format an event as a unicode string.  Optionally, attach timestamp,
+    traceback, and system information.
+
+    The full output format is:
+    C{u"{timeStamp} [{system}] {event}\n{traceback}\n"} where:
+
+        - C{timeStamp} is the event's C{"log_time"} value formatted with
+          the provided C{formatTime} callable.
+
+        - C{system} is the event's C{"log_system"} value, if set, otherwise,
+          the C{"log_namespace"} and C{"log_level"}, joined by a C{u"#"}.  Each
+          defaults to C{u"-"} is not set.
+
+        - C{event} is the event, as formatted by L{formatEvent}.
+
+        - C{traceback} is the traceback if the event contains a
+          C{"log_failure"} key.  In the event the original traceback cannot
+          be formatted, a message indicating the failure will be substituted.
+
+    If the event cannot be formatted, and no traceback exists, an empty string
+    is returned, even if includeSystem or includeTimestamp are true.
+
+    @param event: A logging event.
+    @type event: L{dict}
+
+    @param includeTraceback: If true and a C{"log_failure"} key exists, append
+        a traceback.
+    @type includeTraceback: L{bool}
+
+    @param includeTimestamp: If true include a formatted timestamp before the
+        event.
+    @type includeTimestamp: L{bool}
+
+    @param includeSystem:  If true, include the event's C{"log_system"} value.
+    @type includeSystem: L{bool}
+
+    @param formatTime: A time formatter
+    @type formatTime: L{callable} that takes an C{event} argument and returns
+        a L{unicode}
+
+    @return: A formatted string with specified options.
+    @rtype: L{unicode}
+
+    @since: Twisted NEXT
+    """
+    eventText = _formatEvent(event)
+    if includeTraceback and 'log_failure' in event:
+        f = event['log_failure']
+        traceback = _formatTraceback(f)
+        eventText = u"\n".join((eventText, traceback))
+
+    if not eventText:
+        return eventText
+
+    logText = u""
```

Comment:
Normally I wouldn't be so concerned with a tiny micro-optimization, but given that this has historically been a *very* hot code path in Twisted, I think it might be worthwhile to take some trouble to avoid having to rebuild the format string in its entirety on every single call.

(Let's not block on this; it can easily be done in a follow-up PR.)

Reply:
I'll spend a bit on it, but I won't let it add much delay.

Reply:
@glyph Since you mentioned this is a hot path, it leads me to another question.  There are helper functions `_formatEvent`, `_formatTraceback` used because (to me) they kept the code more readable.  However, the function call has overhead.  Should those be removed in favor of performance?